torchtune/generation/_generation.py

```diff
@@ -69,85 +109,227 @@ def update_stop_tokens_tracker(
 
     Returns:
-        List[List[int]]: collection of lists of generated tokens
+        torch.Tensor: Generated tokens.
     """
     prompt = prompt.view(1, -1) if prompt.ndim == 1 else prompt
-    # convert stop tokens to tensor for easy matching
+
     stop_tokens = (
         torch.tensor(stop_tokens, device=prompt.device) if stop_tokens else None
     )
+
+    if custom_generate_next_token is None:
+        custom_generate_next_token = generate_next_token
+
     bsz, prompt_length = prompt.size()
+    total_response_length = prompt_length + max_generated_tokens
+
     generated_tokens = prompt.clone()
-    # keeps track at a high level if we've already hit a stop token in a sequence so we can early stop
-    stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
-    # everything in stop_token_mask starts as 1s, and we'll set them to 0 for sequences
-    # that already hit a stop token
-    stop_token_mask = torch.ones(
-        (bsz, prompt_length + 1), dtype=torch.int32, device=prompt.device
+    incremental_decoding = model.caches_are_enabled()
+
+    # grab the correct max_seq_len to generate full causal masks/position ids
+    # this is the model's max cache len if incremental decoding, or the sequence
+    # length otherwise
+    max_seq_len = (
+        total_response_length if not incremental_decoding else model.max_seq_len
     )
 
-    if custom_generate_next_token is None:
-        custom_generate_next_token = generate_next_token
+    padding_masks = generated_tokens != pad_id
+    cache_pos = None
 
-    # generate the first tokens conditioned on the prompt
-    input_pos = torch.arange(0, model.max_seq_len, device=prompt.device)
-    tokens = generate_next_token(
+    if not padding_masks.all():
+        # we have padding in the prompt due to varying-length sequences in a batch
+        # extend padding masks out to the correct seq len
+        padding_masks = torch.nn.functional.pad(
+            padding_masks, (0, max_generated_tokens), value=True
+        )
+
+        # generate the full causal mask for the whole padding mask with padding ignored
+        masks = get_causal_mask_from_padding_mask(
+            padding_masks, target_seq_len=max_seq_len
+        )
+
+        # right-shift position IDs to account for padding
+        input_pos = get_position_ids_from_padding_mask(padding_masks)
+        cache_pos = torch.arange(
+            0, total_response_length, device=generated_tokens.device
+        )
+    else:
+        # just use a regular causal mask if there is no padding
+        masks = torch.tril(
+            torch.ones(
+                total_response_length,
+                max_seq_len,
+                dtype=torch.bool,
+                device=prompt.device,
+            )
+        ).unsqueeze(0)
+        input_pos = torch.arange(
+            0, total_response_length, device=generated_tokens.device
+        ).unsqueeze(0)
+
+    del padding_masks
```

Comment:
Just curious - Do we know how much memory this saves?

Reply:
in the best case 1 byte * bsz * model_max_seq_len, so like ~10MB with Llama3.1 and bsz=100. will remove